Sources/Workspace/Workspace.swift

```diff
@@ -1732,19 +1744,24 @@ extension Workspace {
                 targetName: artifact.targetName
             ]
 
-            if let existingArtifact = existingArtifact, case .remote(_, let existingChecksum) = existingArtifact.source {
-                // If we already have an artifact with the same checksum, we don't need to download it again.
-                if artifact.checksum == existingChecksum {
-                    continue
-                }
+            if let existingArtifact = existingArtifact {
+                if case .remote(_, let existingChecksum) = existingArtifact.source {
+                    // If we already have an artifact with the same checksum, we don't need to download it again.
+                    if artifact.checksum == existingChecksum {
+                        continue
+                    }
 
-                // If the checksum is different but the package wasn't updated, this is a security risk.
-                if !addedOrUpdatedPackages.contains(artifact.packageRef) {
-                    diagnostics.emit(.artifactChecksumChanged(targetName: artifact.targetName))
-                    continue
+                    // If the checksum is different but the package wasn't updated, this is a security risk.
+                    if !addedOrUpdatedPackages.contains(artifact.packageRef) {
+                        diagnostics.emit(.artifactChecksumChanged(targetName: artifact.targetName))
+                        continue
+                    }
                 }
 
-                artifactsToRemove.append(existingArtifact)
+                if isAtArtifactsDirectory(existingArtifact) {
+                    // Remove the old extracted artifact, be it local archived or remote one.
```

Comment:
we should make sure this change does not remove local (non-archived) directories, which would delete user's code. if there isn't a unit test for this already, worth adding one

Reply:
this is my only remaining concern ^^

Reply:
This check makes sure that the existing artifact is removed only if it exists in the artifacts directory, which is never the case for local non-archived artifacts. You're right that requirement has to be covered with a test, [added one](https://github.com/apple/swift-package-manager/pull/3746/commits/554e2990d0ac4242f4086af756ad195ef81744e5).